### PR TITLE
Feat: identity precompile

### PR DESCRIPTION
**Pull request by roynalnaruto**
### Description

[_PR description_]

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

[_explanation_]

<hr>

## How to fill a PR description 

Please give a concise description of your PR.

The target readers could be future developers, reviewers, and auditors. By reading your description, they should easily understand the changes proposed in this pull request.

MUST: Reference the issue to resolve

### Single responsability

Is RECOMMENDED to create single responsibility commits, but not mandatory.

Anyway, you MUST enumerate the changes in a unitary way, e.g.

```
This PR contains:
- Cleanup of xxxx, yyyy
- Changed xxxx to yyyy in order to bla bla
- Added xxxx function to ...
- Refactored ....
```

### Design choices

RECOMMENDED to:
- What types of design choices did you face?
- What decisions you have made?
- Any valuable information that could help reviewers to think critically

### CODE REVIEW

**Comment by lispc** [comment]
@silathdiir tesed this branch with testool and found some "precompile revert" cases failed.   I asked steven to have a look and debug these failures.

https://circuit-release.s3.us-west-2.amazonaws.com/testool/default.1684987723.f2c623c.html

**Comment by silathdiir** [comment]
> @silathdiir tesed this branch with testool and found some "precompile revert" cases failed. I asked steven to have a look and debug these failures.
> 
> https://circuit-release.s3.us-west-2.amazonaws.com/testool/default.1684987723.f2c623c.html

After merging https://github.com/scroll-tech/zkevm-circuits/pull/516, the failed testool cases reduce from `123` to `42`.

https://circuit-release.s3.us-west-2.amazonaws.com/testool/default.1685010177.41cabb9.html

I am debugging the remaining failed cases, will try to send another PR. Thanks.

**Comment by lispc** [comment]
during reviewing https://github.com/scroll-tech/zkevm-circuits/pull/517, i realized something.. we should some constraints to check the calldata input for identity is same as the output of identity. Currently, we try to implement this by doing so:

```
// input
cb.copy_table_lookup(
                        cb.curr.state.call_id.expr(),
                        CopyDataType::Memory.expr(),
                        callee_call_id.expr(),
                        CopyDataType::Precompile.expr(), ...
```

```
// ouput 
cb.copy_table_lookup(
                            callee_call_id.expr(),
                            CopyDataType::Precompile.expr(),
                            callee_call_id.expr(),
                            CopyDataType::Memory.expr(), ...
```

But i think it is still not enough?   Did yo constrain in copy circuit that with same call_id, CopyDataType::Precompile for src and dst be equal?   (we should not constraint this, since input/output for other precompiles are not equal)

I am afraid the idea that "BasePrecompile can be IdentityPrecompile" should not be possible.

**Comment by roynalnaruto** [comment]
@lispc oh! that's very well noticed. Not sure how I missed this.

One of the solutions that comes to my mind:
- Instead of `CopyDataType::Precompile`, we should have an `CopyDataType::Precompile(PrecompileCall)`. This also helps us in differentiating between different precompile calls.
- We enable the `rlc_acc` column for those types, for instance, we enable it for `PrecompileCall::Identity`.
- We constrain `rlc_acc_input == rlc_acc_output` (if the call was successful)
- Instead of replacing `IdentityGadget` with `BaseIdentityGadget`, we also pass the `rlc_acc` expressions to the gadget and then apply the necessary operations to the accumulator value. For instance, in the case of `Identity`, we have `rlc_acc_input == rlc_acc_output`. But for SHA256, we can have `SHA256(rlc_acc_input) == rlc_acc_output` (which we can enforce by doing a lookup).

**Comment by lispc** [comment]
i think this new idea will work well

**Comment by roynalnaruto** [comment]
@lispc checking `input_rlc` and `output_rlc` fixed in https://github.com/scroll-tech/zkevm-circuits/pull/514/commits/33497f8e90f467b1d84371c92a087e0718a6318b and https://github.com/scroll-tech/zkevm-circuits/pull/514/commits/92d98ebee38fd2f19a8a0949e412a8b0d929085f

**Comment by lispc** [comment]
the new constrains (input_rlc/len == output_rlc/len) looks correct, while i am considering is it possible to move  

```
                cb.require_equal(
                    "input and output bytes are the same",
                    input_bytes_rlc,
                    output_bytes_rlc,
                );
                cb.require_equal(
                    "input length and precompile return length are the same",
                    cd_length,
                    precompile_return_length,
                );
```

into identity gadgets?  

Later we will implement other precompile, i think writing these input/output related constraints in separate files would be better than putting them all in the precompile_gadget.rs, there will be a long long `cb.condition(address.value_equals(PrecompileCalls::XXX/YYY/ZZZ)..`

File: zkevm-circuits/src/evm_circuit/util/precompile_gadget.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        cd_length: Expression<F>,
+        _rd_offset: Expression<F>,
+        _rd_length: Expression<F>,
+        precompile_return_length: Expression<F>,
+        // input bytes to precompile call.
+        input_bytes_rlc: Expression<F>,
+        // output result from precompile call.
+        output_bytes_rlc: Expression<F>,
+        // returned bytes back to caller.
+        _return_bytes_rlc: Expression<F>,
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/precompile_gadget.rs:36]: do you think we will need `return_bytes_rlc/rd_length` later?   If PrecompileGadget is designed to handle precompile contract computation, then slicing output_bytes to return_bytes by rd_length is the duty of the caller?   So PrecompileGadget does not need to know about `return_bytes_rlc/rd_length`?

**Comment by lispc** [comment]
i think this pr is ready and @silathdiir can you run testool again for this PR to make sure it will not break too many old tests?  If everything goes ok, i will merge this

**Comment by silathdiir** [comment]
@lispc @roynalnaruto may need to fix the new failed `testool` cases (`95` failed in total `+64`):

https://circuit-release.s3.us-west-2.amazonaws.com/testool/default.1686102269.9bb5523.html

I am debugging on these.

**Comment by silathdiir** [comment]
Re-run testool on this branch again, remains `32` failed cases (not all related with precompile).
https://circuit-release.s3.us-west-2.amazonaws.com/testool/default.1686137818.fecc428.html

**Pull request merged**